src/gui/src/clockWidget.cpp

```diff
@@ -0,0 +1,1202 @@
+                                  const QRect& rect,
+                                  sta::Delay min_time,
+                                  sta::Delay max_time,
+                                  sta::Delay mouse_time) const
+{
+  min_time /= unit_scale_;
+  max_time /= unit_scale_;
+  mouse_time /= unit_scale_;
+
+  const sta::Delay time_range = max_time - min_time;
+
+  const int digits = 2;
+
+  const int tick_length = std::min(20, rect.width() / 20);
+
+  auto time_to_y = [min_time, time_range, &rect](sta::Delay time) -> qreal {
+    return (time - min_time) / time_range * static_cast<qreal>(rect.height());
+  };
+
+  auto time_to_string = [this](sta::Delay time, int digits) -> QString {
+    return QString::number(time, 'f', digits) + QString(" ") + unit_suffix_;
+  };
+
+  // draw edge line
+  painter->setPen(QPen(Qt::black, 0));
+  painter->drawLine(rect.bottomLeft(), rect.topLeft());
+
+  const int min_tick_marks = 2;
+  const sta::Delay scale_time = std::pow(10, digits);
+  sta::Delay tick_interval
+      = std::pow(10, std::floor(std::log10(time_range * scale_time)))
+        / scale_time;
+  if (time_range / tick_interval < min_tick_marks) {
+    tick_interval /= 10;
+  }
+  tick_interval = std::max(tick_interval, 1 / scale_time);
+
+  sta::Delay tick = std::floor(min_time / tick_interval) * tick_interval;
+  for (; tick < max_time; tick += tick_interval) {
+    const QPoint tick_start(rect.left(), time_to_y(tick));
+    const QPoint tick_end(tick_start.x() + tick_length, tick_start.y());
+    painter->drawLine(tick_start, tick_end);
+    painter->drawText(tick_end, time_to_string(tick, digits));
+  }
+
+  if (show_mouse_time_tick_) {
+    // draw mouse position
+    painter->setPen(QPen(Qt::gray, 0));
+    const QPoint tick_start(rect.left(), time_to_y(mouse_time));
+    const QPoint tick_end(tick_start.x() + tick_length / 2, tick_start.y());
+    painter->drawLine(tick_start, tick_end);
+
+    const QString mouse_text = time_to_string(mouse_time, digits + 1);
+    const QFontMetrics fm(painter->font());
+
+    QRect text_bounds = fm.boundingRect(mouse_text);
+    text_bounds.moveTo(tick_end.x(),
+                       tick_end.y() - text_bounds.height() + fm.descent());
+    painter->setBrush(
+        QBrush(painter->pen().color().lighter(), Qt::SolidPattern));
+    painter->drawRect(text_bounds);
+
+    painter->setPen(QPen(Qt::black, 0));
+    painter->drawText(tick_end, time_to_string(mouse_time, digits + 1));
+  }
+}
+
+void ClockTreeView::wheelEvent(QWheelEvent* event)
+{
+  const double factor = 1.10;
+
+  const auto anchor = transformationAnchor();
+  setTransformationAnchor(QGraphicsView::AnchorUnderMouse);
+
+  if (event->delta() > 0) {
+    scale(factor, factor);
+  } else if (event->delta() < 0) {
+    scale(1.0 / factor, 1.0 / factor);
+  }
+
+  setTransformationAnchor(anchor);
+}
+
+void ClockTreeView::selectionChanged()
+{
+  for (const auto& sel : scene_->selectedItems()) {
+    QVariant data = sel->data(0);
```

Comment:
The nested-ifs are somewhat ugly.  Is it possible to switch on the type the QVariant holds?

Reply:
I'll see what can be done, I agree I didn't love it either, but it started much smaller :-/ and grew